File: public/api/roll_parser.php
<?php
/**
 * Roll Number Parser for College Attendance System
 * Parses roll number format: YY-[E]PROGRAM-NN
 * Extracts: admission year (YY), shift (E=Evening, none=Morning), program code, sequence number
 */

class RollParser {
    
    /**
     * Parse roll number format: YY-[E]PROGRAM-NN
     * 
     * @param string $roll_number Roll number to parse (e.g., "25-SWT-01", "24-ECSE-02")
     * @return array Parsed roll number information
     */
    public static function parseRollNumber($roll_number) {
        if (empty($roll_number) || !is_string($roll_number)) {
            return [
                'valid' => false,
                'error' => 'Invalid roll number format'
            ];
        }
        
        // Clean the roll number
        $roll_number = strtoupper(trim($roll_number));
        
        // Pattern: YY-[E]PROGRAM-NN
        // YY: 2-digit year (20-99)
        // E: Optional 'E' for Evening shift
        // PROGRAM: 2-4 character program code
        // NN: 2-digit sequence number
        $pattern = '/^(\d{2})-([E]?)([A-Z]{2,4})-(\d{2})$/';
        
        if (!preg_match($pattern, $roll_number, $matches)) {
            return [
                'valid' => false,
                'error' => 'Invalid roll number format. Expected: YY-[E]PROGRAM-NN'
            ];
        }
        
        $year_str = $matches[1];
        $evening_flag = $matches[2];
        $program = $matches[3];
        $sequence = $matches[4];
        
        // Parse year (convert to full year)
        $admission_year = intval($year_str);
        if ($admission_year < 20) {
            $admission_year += 2000;
        } else {
            $admission_year += 2000;
        }
        
        // Determine shift
        $shift = ($evening_flag === 'E') ? 'Evening' : 'Morning';
        
        // Parse sequence number
        $sequence_num = intval($sequence);
        
        return [
            'valid' => true,
            'roll_number' => $roll_number,
            'admission_year' => $admission_year,
            'shift' => $shift,
            'program' => $program,
            'sequence_number' => $sequence_num,
            'is_evening' => $evening_flag === 'E'
        ];
    }
    
    /**
     * Calculate current academic year based on admission year and current date
     * 
     * @param int $admission_year Year of admission
     * @param DateTime|null $current_date Current date. Defaults to now
     * @return int Current academic year (1-4)
     */
    public static function getStudentYear($admission_year, $current_date = null) {
        if ($current_date === null) {
            $current_date = new DateTime();
        }
        
        $current_year = intval($current_date->format('Y'));
        $current_month = intval($current_date->format('n'));
        
        // Academic year starts in September
        if ($current_month >= 9) {
            $academic_year = $current_year;
        } else {
            $academic_year = $current_year - 1;
        }
        
        // Calculate year of study
        $year_of_study = $academic_year - $admission_year + 1;
        
        // Cap at 4 years (graduation)
        return min(max($year_of_study, 1), 4);
    }
    
    /**
     * Extract shift from roll number
     * 
     * @param string $roll_number Roll number to parse
     * @return string 'Morning' or 'Evening'
     */
    public static function getShift($roll_number) {
        $parsed = self::parseRollNumber($roll_number);
        if (!$parsed['valid']) {
            return 'Morning'; // Default to morning
        }
        
        return $parsed['shift'];
    }
    
    /**
     * Extract program code from roll number
     * 
     * @param string $roll_number Roll number to parse
     * @return string Program code
     */
    public static function getProgram($roll_number) {
        $parsed = self::parseRollNumber($roll_number);
        if (!$parsed['valid']) {
            return 'UNKNOWN';
        }
        
        return $parsed['program'];
    }
    
    /**
     * Validate roll number format
     * 
     * @param string $roll_number Roll number to validate
     * @return bool True if valid, False otherwise
     */
    public static function validateRollNumber($roll_number) {
        $parsed = self::parseRollNumber($roll_number);
        return $parsed['valid'];
    }
    
    /**
     * Get comprehensive academic year information for a roll number
     * 
     * @param string $roll_number Roll number to analyze
     * @param DateTime|null $current_date Current date. Defaults to now
     * @return array Academic year information
     */
    public static function getAcademicYearInfo($roll_number, $current_date = null) {
        $parsed = self::parseRollNumber($roll_number);
        
        if (!$parsed['valid']) {
            return [
                'valid' => false,
                'error' => $parsed['error']
            ];
        }
        
        $current_year = self::getStudentYear($parsed['admission_year'], $current_date);
        
        return [
            'valid' => true,
            'roll_number' => $roll_number,
            'admission_year' => $parsed['admission_year'],
            'current_year' => $current_year,
            'shift' => $parsed['shift'],
            'program' => $parsed['program'],
            'sequence_number' => $parsed['sequence_number'],
            'is_evening' => $parsed['is_evening'],
            'is_graduated' => $current_year > 4,
            'years_remaining' => max(0, 4 - $current_year)
        ];
    }
    
    /**
     * Get batch information for a given admission year
     * 
     * @param int $admission_year Admission year
     * @return array Batch information
     */
    public static function getBatchInfo($admission_year) {
        $current_date = new DateTime();
        $current_year = self::getStudentYear($admission_year, $current_date);
        
        return [
            'admission_year' => $admission_year,
            'current_year' => $current_year,
            'batch_name' => "Batch {$admission_year}",
            'is_graduated' => $current_year > 4,
            'years_remaining' => max(0, 4 - $current_year)
        ];
    }
    
    /**
     * Get program information based on program code
     * 
     * @param string $program_code Program code (e.g., 'SWT', 'ECSE', 'CS')
     * @return array Program information
     */
    public static function getProgramInfo($program_code) {
        $program_mapping = [
            'SWT' => ['name' => 'Software Technology', 'duration' => 4, 'type' => 'Bachelor'],
            'ECSE' => ['name' => 'Electrical & Computer Systems Engineering', 'duration' => 4, 'type' => 'Bachelor'],
            'CS' => ['name' => 'Computer Science', 'duration' => 4, 'type' => 'Bachelor'],
            'IT' => ['name' => 'Information Technology', 'duration' => 4, 'type' => 'Bachelor'],
            'SE' => ['name' => 'Software Engineering', 'duration' => 4, 'type' => 'Bachelor'],
            'CE' => ['name' => 'Computer Engineering', 'duration' => 4, 'type' => 'Bachelor'],
            'EE' => ['name' => 'Electrical Engineering', 'duration' => 4, 'type' => 'Bachelor'],
            'ME' => ['name' => 'Mechanical Engineering', 'duration' => 4, 'type' => 'Bachelor'],
            'CIVIL' => ['name' => 'Civil Engineering', 'duration' => 4, 'type' => 'Bachelor'],
            'MATH' => ['name' => 'Mathematics', 'duration' => 4, 'type' => 'Bachelor'],
            'PHYS' => ['name' => 'Physics', 'duration' => 4, 'type' => 'Bachelor'],
            'CHEM' => ['name' => 'Chemistry', 'duration' => 4, 'type' => 'Bachelor'],
            'BIO' => ['name' => 'Biology', 'duration' => 4, 'type' => 'Bachelor'],
            'BUS' => ['name' => 'Business Administration', 'duration' => 4, 'type' => 'Bachelor'],
            'ECON' => ['name' => 'Economics', 'duration' => 4, 'type' => 'Bachelor'],
            'PSY' => ['name' => 'Psychology', 'duration' => 4, 'type' => 'Bachelor'],
            'ENG' => ['name' => 'English', 'duration' => 4, 'type' => 'Bachelor'],
            'HIST' => ['name' => 'History', 'duration' => 4, 'type' => 'Bachelor'],
            'ART' => ['name' => 'Art', 'duration' => 4, 'type' => 'Bachelor'],
            'MUS' => ['name' => 'Music', 'duration' => 4, 'type' => 'Bachelor']
        ];
        
        $program_code = strtoupper($program_code);
        $program_info = isset($program_mapping[$program_code]) 
            ? $program_mapping[$program_code] 
            : [
                'name' => "Program {$program_code}",
                'duration' => 4,
                'type' => 'Bachelor'
            ];
        
        return [
            'code' => $program_code,
            'name' => $program_info['name'],
            'duration' => $program_info['duration'],
            'type' => $program_info['type']
        ];
    }
    
    /**
     * Parse multiple roll numbers and return summary statistics
     * 
     * @param array $roll_numbers List of roll numbers to parse
     * @return array Summary statistics
     */
    public static function parseMultipleRollNumbers($roll_numbers) {
        $valid_rolls = [];
        $invalid_rolls = [];
        
        foreach ($roll_numbers as $roll) {
            $parsed = self::parseRollNumber($roll);
            if ($parsed['valid']) {
                $valid_rolls[] = $parsed;
            } else {
                $invalid_rolls[] = [
                    'roll_number' => $roll,
                    'error' => $parsed['error']
                ];
            }
        }
        
        // Analyze valid rolls
        $programs = [];
        $shifts = ['Morning' => 0, 'Evening' => 0];
        $years = [];
        
        foreach ($valid_rolls as $roll) {
            // Count programs
            $program = $roll['program'];
            $programs[$program] = isset($programs[$program]) ? $programs[$program] + 1 : 1;
            
            // Count shifts
            $shifts[$roll['shift']]++;
            
            // Count years
            $year = $roll['admission_year'];
            $years[$year] = isset($years[$year]) ? $years[$year] + 1 : 1;
        }
        
        return [
            'total_rolls' => count($roll_numbers),
            'valid_rolls' => count($valid_rolls),
            'invalid_rolls' => count($invalid_rolls),
            'valid_roll_data' => $valid_rolls,
            'invalid_roll_data' => $invalid_rolls,
            'programs' => $programs,
            'shifts' => $shifts,
            'years' => $years
        ];
    }
}

// Example usage and testing
if (basename(__FILE__) == basename($_SERVER['SCRIPT_NAME'])) {
    // Test cases
    $test_rolls = [
        "25-SWT-01",      // Valid morning shift
        "24-ECSE-02",     // Valid morning shift
        "23-ECS-03",      // Valid morning shift
        "25-ESWT-04",     // Valid evening shift
        "26-CS-05",       // Valid morning shift
        "invalid-roll",   // Invalid
        "25-SWT",         // Invalid (missing sequence)
        "25-SWT-1",       // Invalid (single digit sequence)
        "5-SWT-01",       // Valid (2005)
        "99-SWT-01",      // Valid (2099)
    ];
    
    echo "Roll Number Parser Test Results\n";
    echo str_repeat("=", 50) . "\n";
    
    foreach ($test_rolls as $roll) {
        $result = RollParser::parseRollNumber($roll);
        echo "Roll: {$roll}\n";
        if ($result['valid']) {
            echo "  Valid: Yes\n";
            echo "  Admission Year: {$result['admission_year']}\n";
            echo "  Shift: {$result['shift']}\n";
            echo "  Program: {$result['program']}\n";
            echo "  Sequence: {$result['sequence_number']}\n";
            
            // Get academic year info
            $academic_info = RollParser::getAcademicYearInfo($roll);
            if ($academic_info['valid']) {
                echo "  Current Year: {$academic_info['current_year']}\n";
                echo "  Graduated: " . ($academic_info['is_graduated'] ? 'Yes' : 'No') . "\n";
            }
        } else {
            echo "  Valid: No - {$result['error']}\n";
        }
        echo "\n";
    }
    
    // Test batch analysis
    echo "Batch Analysis\n";
    echo str_repeat("=", 30) . "\n";
    $batch_info = RollParser::getBatchInfo(2024);
    echo "Batch 2024: Year {$batch_info['current_year']}, Graduated: " . ($batch_info['is_graduated'] ? 'Yes' : 'No') . "\n";
    
    // Test program info
    echo "\nProgram Information\n";
    echo str_repeat("=", 30) . "\n";
    $program_info = RollParser::getProgramInfo('SWT');
    echo "SWT: {$program_info['name']} ({$program_info['type']})\n";
    
    // Test multiple roll parsing
    echo "\nMultiple Roll Analysis\n";
    echo str_repeat("=", 30) . "\n";
    $analysis = RollParser::parseMultipleRollNumbers(array_slice($test_rolls, 0, 5)); // Only valid ones
    echo "Total: {$analysis['total_rolls']}, Valid: {$analysis['valid_rolls']}, Invalid: {$analysis['invalid_rolls']}\n";
    echo "Programs: " . json_encode($analysis['programs']) . "\n";
    echo "Shifts: " . json_encode($analysis['shifts']) . "\n";
    echo "Years: " . json_encode($analysis['years']) . "\n";
}
?>

File: python/roll_parser.py
import re
from datetime import datetime

def parse_roll_number(roll_number):
    """
    Parse roll number to extract academic information with enhanced validation.
    
    Expected format: YY-PPP-RR
    - YY: Admission year (last two digits)
    - PPP: Program code (e.g., SWT, MIT, BSC)
    - RR: Roll sequence number
    
    Example: 24-SWT-01 -> Admission Year: 2024, Program: SWT, Roll: 01
    """
    try:
        # Clean the input
        roll_number = str(roll_number).strip().upper()
        
        # Enhanced pattern matching for various roll number formats
        patterns = [
            r'^(\d{2})[-_]?([A-Z]{2,4})[-_]?(\d{2,3})$',  # 24-SWT-01 or 24_SWT_01
            r'^([A-Z]{2,4})[-_]?(\d{2})[-_]?(\d{2,3})$',  # SWT-24-01
            r'^(\d{2})([A-Z]{2,4})(\d{2,3})$',           # 24SWT01
        ]
        
        for pattern in patterns:
            match = re.match(pattern, roll_number)
            if match:
                groups = match.groups()
                
                if len(groups) == 3:
                    # Determine format based on group content
                    if groups[0].isdigit() and len(groups[0]) == 2 and groups[1].isalpha():
                        # Format: YY-PPP-RR
                        admission_year_short = int(groups[0])
                        program_code = groups[1]
                        roll_sequence = groups[2]
                    elif groups[0].isalpha() and groups[1].isdigit() and len(groups[1]) == 2:
                        # Format: PPP-YY-RR
                        program_code = groups[0]
                        admission_year_short = int(groups[1])
                        roll_sequence = groups[2]
                    else:
                        continue
                    
                    # Calculate full admission year
                    current_year = datetime.now().year
                    current_century = current_year // 100 * 100
                    admission_year = current_century + admission_year_short
                    
                    # Adjust for future/past years
                    if admission_year > current_year + 5:
                        admission_year -= 100
                    elif admission_year < current_year - 10:
                        admission_year += 100
                    
                    return {
                        'admission_year': admission_year,
                        'program_code': program_code,
                        'roll_sequence': roll_sequence,
                        'original_roll': roll_number
                    }
        
        # If no pattern matched, try to extract basic information
        print(f"⚠️ Could not parse roll number format: {roll_number}")
        return {
            'admission_year': datetime.now().year,
            'program_code': 'UNK',
            'roll_sequence': '01',
            'original_roll': roll_number
        }
        
    except Exception as e:
        print(f"❌ Error parsing roll number {roll_number}: {e}")
        return {
            'admission_year': datetime.now().year,
            'program_code': 'UNK',
            'roll_sequence': '01',
            'original_roll': roll_number
        }

def get_shift(roll_number):
    """Determine student shift based on roll number with enhanced logic."""
    try:
        parsed = parse_roll_number(roll_number)
        program = parsed['program_code']
        
        # Enhanced shift determination logic
        shift_mapping = {
            'SWT': 'Morning',
            'MIT': 'Evening', 
            'BSC': 'Morning',
            'MSC': 'Evening',
            'PHD': 'Evening'
        }
        
        return shift_mapping.get(program, 'Morning')  # Default to Morning
        
    except Exception as e:
        print(f"❌ Error determining shift for {roll_number}: {e}")
        return 'Morning'

def get_program(roll_number):
    """Get full program name from roll number with enhanced mapping."""
    try:
        parsed = parse_roll_number(roll_number)
        program_code = parsed['program_code']
        
        program_mapping = {
            'SWT': 'Software Technology',
            'MIT': 'Master of IT',
            'BSC': 'Bachelor of Science',
            'MSC': 'Master of Science',
            'PHD': 'Doctor of Philosophy',
            'CS': 'Computer Science',
            'IT': 'Information Technology',
            'SE': 'Software Engineering'
        }
        
        return program_mapping.get(program_code, f"Program {program_code}")
        
    except Exception as e:
        print(f"❌ Error determining program for {roll_number}: {e}")
        return 'Unknown Program'

def get_academic_year_info(roll_number):
    """Calculate current academic year and admission year with enhanced logic."""
    try:
        parsed = parse_roll_number(roll_number)
        admission_year = parsed['admission_year']
        current_year = datetime.now().year
        current_month = datetime.now().month
        
        # Academic year calculation (starts in August)
        if current_month >= 8:  # August to December
            academic_year = current_year - admission_year + 1
        else:  # January to July
            academic_year = current_year - admission_year
        
        # Ensure academic year is reasonable
        if academic_year < 1:
            academic_year = 1
        elif academic_year > 10:  # Maximum reasonable years
            academic_year = 10
        
        return academic_year, admission_year
        
    except Exception as e:
        print(f"❌ Error calculating academic year for {roll_number}: {e}")
        return 1, datetime.now().year

def validate_roll_number(roll_number):
    """Validate roll number format with enhanced checks."""
    try:
        parsed = parse_roll_number(roll_number)
        
        # Check if parsing was successful
        if parsed['program_code'] == 'UNK':
            return False
        
        # Additional validation checks
        current_year = datetime.now().year
        admission_year = parsed['admission_year']
        
        # Check if admission year is reasonable
        if admission_year < 2000 or admission_year > current_year + 1:
            return False
        
        # Check roll sequence
        roll_seq = parsed['roll_sequence']
        if not roll_seq.isdigit() or int(roll_seq) <= 0:
            return False
        
        return True
        
    except Exception as e:
        print(f"❌ Error validating roll number {roll_number}: {e}")
        return False

if __name__ == "__main__":
    # Test the roll parser
    test_rolls = ['24-SWT-01', '23-MIT-15', 'SWT-24-01', '25BSC05', 'INVALID']
    
    print("Roll Parser Test:")
    for roll in test_rolls:
        print(f"\nRoll: {roll}")
        parsed = parse_roll_number(roll)
        print(f"Parsed: {parsed}")
        print(f"Shift: {get_shift(roll)}")
        print(f"Program: {get_program(roll)}")
        print(f"Academic Year: {get_academic_year_info(roll)}")
        print(f"Valid: {validate_roll_number(roll)}")